File: Client_Start/Assets/Scripts/Libs/EventHandle/EventDispatch.cs
﻿using LuaInterface;
using System;

namespace SDK.Lib
{
    /**
     * @brief 事件分发，之分发一类事件，不同类型的事件使用不同的事件分发
     * @brief 注意，事件分发缺点就是，可能被调用的对象已经释放，但是没有清掉事件处理器，结果造成空指针
     */
    public class EventDispatch : DelayHandleMgrBase
    {
        protected int mEventId;
        protected MList<EventDispatchFunctionObject> mHandleList;
        protected int mUniqueId;       // 唯一 Id ，调试使用
        protected LuaCSBridgeDispatch mLuaCSBridgeDispatch;

        public EventDispatch(int eventId_ = 0)
        {
            this.mEventId = eventId_;
            this.mHandleList = new MList<EventDispatchFunctionObject>();
        }

        protected MList<EventDispatchFunctionObject> handleList
        {
            get
            {
                return this.mHandleList;
            }
        }

        public int uniqueId
        {
            get
            {
                return this.mUniqueId;
            }
            set
            {
                this.mUniqueId = value;
                this.mHandleList.uniqueId = this.mUniqueId;
            }
        }

        public LuaCSBridgeDispatch luaCSBridgeDispatch
        {
            get
            {
                return this.mLuaCSBridgeDispatch;
            }
            set
            {
                this.mLuaCSBridgeDispatch = value;
            }
        }

        // 相同的函数只能增加一次，Lua ，Python 这些语言不支持同时存在几个相同名字的函数，只支持参数可以赋值，因此不单独提供同一个名字不同参数的接口了
        virtual public void addEventHandle(ICalleeObject pThis, MAction<IDispatchObject> handle, LuaTable luaTable = null, LuaFunction luaFunction = null)
        {
            if (null != pThis || null != handle || null != luaTable || null != luaFunction)
            {
                EventDispatchFunctionObject funcObject = new EventDispatchFunctionObject();
                if (null != handle)
                {
                    funcObject.setFuncObject(pThis, handle);
                }
                if(null != luaTable || null != luaFunction)
                {
                    funcObject.setLuaFunctor(luaTable, luaFunction);
                }
                addObject(funcObject);
            }
            else
            {
                Ctx.mInstance.mLogSys.log("Event Handle is null");
            }
        }

        public void removeEventHandle(ICalleeObject pThis, MAction<IDispatchObject> handle, LuaTable luaTable = null, LuaFunction luaFunction = null)
        {
            int idx = 0;
            int elemLen = 0;
            elemLen = this.mHandleList.Count();
            while (idx < elemLen)
            {
                if (this.mHandleList[idx].isEqual(pThis, handle, luaTable, luaFunction))
                {
                    break;
                }

                idx += 1;
            }
            if (idx < this.mHandleList.Count())
            {
                removeObject(this.mHandleList[idx]);
            }
            else
            {
                Ctx.mInstance.mLogSys.log("Event Handle not exist");
            }
        }

        override protected void addObject(IDelayHandleItem delayObject, float priority = 0.0f)
        {
            if (bInDepth())
            {
                base.addObject(delayObject, priority);
            }
            else
            {
                // 这个判断说明相同的函数只能加一次，但是如果不同资源使用相同的回调函数就会有问题，但是这个判断可以保证只添加一次函数，值得，因此不同资源需要不同回调函数
                this.mHandleList.Add(delayObject as EventDispatchFunctionObject);
            }
        }

        override protected void removeObject(IDelayHandleItem delayObject)
        {
            if (bInDepth())
            {
                base.removeObject(delayObject);
            }
            else
            {
                if (!this.mHandleList.Remove(delayObject as EventDispatchFunctionObject))
                {
                    Ctx.mInstance.mLogSys.log("Event Handle not exist");
                }
            }
        }

        virtual public void dispatchEvent(IDispatchObject dispatchObject)
        {
            //try
            //{
                incDepth();

                foreach (EventDispatchFunctionObject handle in this.mHandleList.list())
                {
                    if (!handle.mIsClientDispose)
                    {
                        handle.call(dispatchObject);
                    }
                }

                if (this.mLuaCSBridgeDispatch != null)
                {
                this.mLuaCSBridgeDispatch.handleGlobalEvent(this.mEventId, dispatchObject);
                }

                decDepth();
            //}
            //catch (Exception ex)
            //{
            //    Ctx.mInstance.mLogSys.catchLog(ex.ToString());
            //}
        }

        public void clearEventHandle()
        {
            if (bInDepth())
            {
                foreach (EventDispatchFunctionObject item in this.mHandleList.list())
                {
                    removeObject(item);
                }
            }
            else
            {
                this.mHandleList.Clear();
            }
        }

        // 这个判断说明相同的函数只能加一次，但是如果不同资源使用相同的回调函数就会有问题，但是这个判断可以保证只添加一次函数，值得，因此不同资源需要不同回调函数
        public bool existEventHandle(ICalleeObject pThis, MAction<IDispatchObject> handle, LuaTable luaTable = null, LuaFunction luaFunction = null)
        {
            bool bFinded = false;
            foreach (EventDispatchFunctionObject item in this.mHandleList.list())
            {
                if (item.isEqual(pThis, handle, luaTable, luaFunction))
                {
                    bFinded = true;
                    break;
                }
            }

            return bFinded;
        }

        public void copyFrom(EventDispatch rhv)
        {
            foreach(EventDispatchFunctionObject handle in rhv.handleList.list())
            {
                this.mHandleList.Add(handle);
            }
        }

        public bool hasEventHandle()
        {
            return this.mHandleList.Count() > 0;
        }
    }
}

File: Bak/scripts/MyLua/Libs/EventHandle/EventDispatch.lua
MLoader("MyLua.Libs.Core.GlobalNS");
MLoader("MyLua.Libs.Core.Class");
MLoader("MyLua.Libs.DataStruct.MList");
MLoader("MyLua.Libs.DelayHandle.DelayPriorityHandleMgrBase");
MLoader("MyLua.Libs.EventHandle.EventDispatchFunctionObject");

--[[
    @brief 事件分发器
]]

local M = GlobalNS.Class(GlobalNS.DelayPriorityHandleMgrBase);
M.clsName = "EventDispatch";
GlobalNS[M.clsName] = M;

function M:ctor(eventId_)
    self.mEventId = eventId_;
    self.mHandleList = GlobalNS.new(GlobalNS.MList);
    self.mUniqueId = 0;       -- 唯一 Id ，调试使用
end

function M:dtor()

end

function M:getHandleList()
    return self.mHandleList;
end

function M:getUniqueId()
    return self.mUniqueId;
end

function M:setUniqueId(value)
    self.mUniqueId = value;
    --self.mHandleList.uniqueId = mUniqueId;
end

function M:addEventHandle(pThis, handle)
	if (nil ~= handle) then
		local funcObject = GlobalNS.new(GlobalNS.EventDispatchFunctionObject);
		funcObject:setFuncObject(pThis, handle);
		self:addObject(funcObject);
	else
		if(MacroDef.ENABLE_LOG) then
			GCtx.mLogSys:log("EventDispatch::addEventHandle, error, add failed", GlobalNS.LogTypeId.eLogEventHandle);
		end
	end
end

function M:removeEventHandle(handle, pThis)
    local idx = 0;
	local listLen = self.mHandleList:Count();
	--for idx = 0, self.mHandleList:Count() - 1, 1 do
    while(idx < listLen) do
        if (self.mHandleList:get(idx):isEqual(handle, pThis)) then
            break;
        end
		
		idx = idx + 1;
    end
	
    if (idx < self.mHandleList:Count()) then
        self:removeObject(self.mHandleList:get(idx));
    else
		if(MacroDef.ENABLE_LOG) then
			GCtx.mLogSys:log("EventDispatch::removeEventHandle, error, remove failed", GlobalNS.LogTypeId.eLogEventHandle);
		end
    end
end

function M:addObject(delayObject, priority)
    if (self:isInDepth()) then
        M.super.addObject(self, delayObject, priority); -- super 使用需要自己填充 Self 参数
    else
        -- 这个判断说明相同的函数只能加一次，但是如果不同资源使用相同的回调函数就会有问题，但是这个判断可以保证只添加一次函数，值得，因此不同资源需要不同回调函数
        self.mHandleList:Add(delayObject);
    end
end

function M:removeObject(delayObject)
    if (self:isInDepth()) then
        M.super.removeObject(self, delayObject);
    else
        if (self.mHandleList:Remove(delayObject) == false) then
            -- 日志
        end
    end
end

function M:dispatchEvent(dispatchObject)
    self:incDepth();

    for _, handle in ipairs(self.mHandleList:list()) do
        if (handle.mIsClientDispose == false) then
            handle:call(dispatchObject);
        end
    end

    self:decDepth();
end

function M:clearEventHandle()
    if (self:isInDepth()) then
        for _, item in ipairs(self.mHandleList:list()) do
            self:removeObject(item);
        end
    else
        self.mHandleList:Clear();
    end
end

-- 这个判断说明相同的函数只能加一次，但是如果不同资源使用相同的回调函数就会有问题，但是这个判断可以保证只添加一次函数，值得，因此不同资源需要不同回调函数
function M:isExistEventHandle(pThis, handle)
    local isFinded = false;
	
    --for _, item in ipairs(self.mHandleList:list()) do
	local index = 0;
	local listLen = self.mHandleList:count();
	local item = nil;
	
	while(index < listLen) do
		item = self.mHandleList:get(index);
		
        if (item:isEqual(pThis, handle)) then
            isFinded = true;
            break;
        end
		
		index = index + 1;
    end

    return isFinded;
end

function M:copyFrom(rhv)
    --for _, handle in ipairs(rhv.handleList:list()) do
	
	local index = 0;
	local listLen = rhv.handleList:count();
	local handle = nil;
	
	while(index < listLen) do
		handle = rhv.handleList:get(index);
        self.mHandleList:Add(handle);
		
		index = index + 1;
    end
end

return M;